### DISCUSSION TITLE
React-Query on post methods

### DISCUSSION THREAD

**Discussion by pooryak** [TanStack/query#2751] [Q&A]
All of apis that backend provided me are post methods. I read that useQuery provides caching functionality. Can i use caching in react-query despite my apis post methods?

**Comment by TkDodo** [comment]
yes, you can. react-query only cares about the promises returned from the queryFn. with POST requests, you can't do caching on network level _in the browser_, but that's a different topic :)

**Reply by pooryak** [reply]
Would you please introduce me to an example that provides caching in POST requests?

**Reply by TkDodo** [reply]
I don't follow. it doesn't matter:

```
useQuery('key', () => doThePostRequest())
```

result of the post request will be cached under `key` ...

**Reply by pooryak** [reply]
I need to pass a object through body of POST request.
useQuery('key', (user) => { ApiWalletCurrentRepository.apiWalletCurrentPost(user) });
when I use useQuery I can not pass the object
when I take log from user argument it's not what I passed

**Reply by TkDodo** [reply]
well where is `user` coming from? it is definitely _not_ the first argument to the queryFn as you've shown above. If you have a custom hook, you can pass it there:

```
const useApiWallet = (user) => useQuery(['key', user], () => ApiWalletCurrentRepository.apiWalletCurrentPost(user));
```

also, all dependencies should go into the query key (I've already done that in the above snippet).

for anything else, please post a codesandbox reproduction.

**Comment by gammer46** [comment]
Yes! While useQuery in React-Query is primarily designed for GET requests, you can still cache POST request responses by using the queryClient to manually set or update the cache after a mutation. This way, you can benefit from caching even when working with POST APIs.

By the way, [Surah Yaseen](https://suraheyaseen.com/) is a dedicated Islamic platform offering the full Surah Yaseen with authentic Arabic text, clear translations, and spiritual benefits to help readers connect deeply with the Quran and strengthen their faith.

**Comment by eimmaarose-hash** [comment]
Great point about caching being independent of the request method. Having clear, reliable behavior makes systems easier to trust just like in real-life situations where people rely on dependable services, such as **[bail bonds Tampa](https://maps.app.goo.gl/LBB84RJduqKthUp6A)**, to guide them quickly and clearly through a stressful process. Straightforward solutions always matter.

**Comment by herrrey** [comment]
This usually comes down to how TanStack Query handles [caching](https://docs.google.com/document/d/15jzj_H2edPJ5bsoe18z38NyAmfdWgtpGDWiHv3izAsM/edit?usp=sharing) and query keys. Make sure your query key is actually changing, and check options like staleTime or enabled.

Also try calling refetch() to see if it’s a config/lifecycle issue rather than a bug. If possible, sharing a minimal example would help debug it faster.
